### PR TITLE
feat: part.type for easy type narrowing

### PR DESCRIPTION
**Pull request by nadhifikbarw**
Easier type narrowing as discussed #144

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by nadhifikbarw** [comment]
Before  
image

After  
image

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]
It looks better.

File: test/index.test-d.ts [hunk 1]
```diff
@@ -55,7 +55,8 @@ const runServer = async () => {
   app.post('/', async (req, reply) => {
     const data = await req.file()
     if (data == null) throw new Error('missing file')
-
+    
+    expectAssignable<string>(data.type)
```
Comment (Uzlopak) [inline test/index.test-d.ts:59]:
```suggestion
    expectType<'file'>(data.type)
```

File: test/index.test-d.ts [hunk 2]
```diff
@@ -88,7 +89,7 @@ const runServer = async () => {
     expectType<string>(req.body.foo.value);
```
Comment (Uzlopak) [inline test/index.test-d.ts:89]:
```suggestion
    expectType<'field'>(req.body.foo.type)
    expectType<string>(req.body.foo.value);
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]
Just some nitpicking on the typings tests, other than LGTM. Nice work.

**Comment by nadhifikbarw** [comment]
@Uzlopak thanks! that was my initial approach, I thought it won't work since my IDE showing error, but it was only my cache issue.

**Review by Uzlopak — APPROVED** [review]
LGTM

**Pull request merged**